tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py

```diff
@@ -0,0 +1,353 @@
+
+_GridRNNDimension = namedtuple('_GridRNNDimension', ['idx', 'is_input', 'is_output', 'is_priority', 'non_recurrent_fn'])
+
+_GridRNNConfig = namedtuple('_GridRNNConfig', ['num_dims', 'dims',
+                                               'inputs', 'outputs', 'recurrents',
+                                               'priority', 'non_priority', 'tied', 'num_units'])
+
+
+def _parse_rnn_config(num_dims, ls_input_dims, ls_output_dims, ls_priority_dims, ls_non_recurrent_dims,
+                      non_recurrent_fn, tied, num_units):
+  def check_dim_list(ls):
+    if ls is None:
+      ls = []
+    if not isinstance(ls, (list, tuple)):
+      ls = [ls]
+    ls = sorted(set(ls))
+    if any(_ < 0 or _ >= num_dims for _ in ls):
+      raise ValueError('Invalid dims: {}. Must be in [0, {})'.format(ls, num_dims))
+    return ls
+
+  input_dims = check_dim_list(ls_input_dims)
+  output_dims = check_dim_list(ls_output_dims)
+  priority_dims = check_dim_list(ls_priority_dims)
+  non_recurrent_dims = check_dim_list(ls_non_recurrent_dims)
+
+  rnn_dims = []
+  for i in range(num_dims):
+    rnn_dims.append(_GridRNNDimension(idx=i, is_input=(i in input_dims), is_output=(i in output_dims),
+                                      is_priority=(i in priority_dims),
+                                      non_recurrent_fn=non_recurrent_fn if i in non_recurrent_dims else None))
+  return _GridRNNConfig(num_dims=num_dims, dims=rnn_dims, inputs=input_dims, outputs=output_dims,
+                        recurrents=[x for x in range(num_dims) if x not in non_recurrent_dims],
+                        priority=priority_dims,
+                        non_priority=[x for x in range(num_dims) if x not in priority_dims],
+                        tied=tied, num_units=num_units)
+
+
+def _get_first(*args):
+  return next((i for i in args if i is not None), None)
+
+
+def _propagate(dim_indices, conf, cell, c_prev, m_prev, new_output, new_state, first_call):
+  """
+  """
+  if len(dim_indices) == 0:
+    return
+
+  if conf.num_dims > 1:
+    ls_cell_inputs = [None] * (conf.num_dims - 1)
+    for d in conf.dims[:-1]:
+      ls_cell_inputs[d.idx] = _get_first(new_output[d.idx], m_prev[d.idx])
+    cell_inputs = array_ops.concat(1, ls_cell_inputs)
+  else:
+    cell_inputs = array_ops.zeros([m_prev[0].get_shape().as_list()[0], 0], m_prev[0].dtype)
+
+  last_dim_output = _get_first(new_output[-1], m_prev[-1])
+
+  for i in dim_indices:
+    d = conf.dims[i]
+    if d.non_recurrent_fn:
+      linear_args = [cell_inputs, last_dim_output] if conf.num_dims > 1 else last_dim_output
+      if conf.tied:
+        with vs.variable_scope('non_recurrent', reuse=not(first_call and i == dim_indices[0])):
+          new_output[d.idx] = d.non_recurrent_fn(rnn_cell.linear(args=linear_args,
+                                                                 output_size=conf.num_units, bias=True))
+      else:
+          new_output[d.idx] = d.non_recurrent_fn(rnn_cell.linear(args=linear_args,
```

Comment:
can you use the contrib.layers.fully_connected here instead?  we're trying to get rid of rnn_cell.linear with fire.